src/System.Private.CoreLib/shared/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -16,16 +16,23 @@ internal partial class ClrThreadPool
         private WaitThreadNode _waitThreadsHead;
         private WaitThreadNode _waitThreadsTail;
 
-        private LowLevelLock _waitThreadLock = new LowLevelLock();
+#if CORERT
+        private readonly Lock _waitThreadLock = new Lock();
+#else
+        private object _waitThreadLock = new object();
```

Comment:
I considered it as an option. It would likely result in some counter-part like `[LowLevel]LockHolder` that can be used in `using (new LockHolder(foo))`. There's only few places so #if's seemed like an option for now, but I am going to revisit the options before dropping the "WIP" tag.

Reply:
I'm specifically going to evaluate the performance of `LowLevelLock` (only used in two places; has the Windows implementation that is not used anywhere else) vs. `Lock` (used to implement Monitor, so should be pretty optimized) vs. `lock (obj)` (portable, but goes through one extra hoop through ObjectHeader/SyncTable; the impact on the specific places might be negligible) to see whether there is any merit preferring one over the others.

Reply:
Implementation-wise, `Lock` in CoreRT is very roughly based on the old implementation of `Monitor` with some major differences. It is quite likely that currently it is not as good as `Monitor`'s current implementation in CoreCLR. Eventually that implementation would also have to be ported to CoreRT. In any case, see my other comment for my suggestions.

Reply:
> `LowLevelLock` (only used in two places; has the Windows PInvoke-heavy implementation that is not used anywhere else)

`LowLevelLock` does some (tiny) bit of spin-waiting and where it is used for the most part the locks are typically uncontended, so it's not really p/invoke-heavy in practice (if it turns out to be, the spin-waiting strategy could be improved to fix that).

Reply:
The motivation behind this change was two-fold:
- Avoid bringing `LowLevelLock`/`LowLevelMonitor` to shared CoreLib (and all the dependent code).
- Use the same primitive as the `lock (...)` pattern. The premise was that this is what user code gets to use and it should be the one that is optimized better.

The problem is that it violates the condition that it is non-interruptible. Another problem is that I grossly underestimated the dependencies from `LowLevelLifoSemaphore`, so in my revised plan I would bring the dependencies to shared code anyway and this change would become unnecessary.